Convert cql-to-elm preprocessor Java code to Kotlin

antvaset opened the pull request:
Using IntelliJ's "Convert Java File to Kotlin File" with manual fixes.

All classes under `org.cqframework.cql.cql2elm.preprocessor` are converted.

JPercival commented on Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/preprocessor/CqlPreprocessorElmCommonVisitor.kt:
```diff
@@ -0,0 +1,878 @@
+    @JvmField var includeDeprecatedElements = false
+
+    init {
+        this.libraryBuilder = Objects.requireNonNull(libraryBuilder, "libraryBuilder required")
```
this initialization is no longer needed, since Kotlin is non-nullable by default

antvaset replied:
:+1: I've removed this line because `libraryBuilder` is declared as `LibraryBuilder`

JPercival commented on Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/preprocessor/CqlPreprocessor.kt:
```diff
@@ -0,0 +1,316 @@
+
+    override fun visitQualifiedIdentifier(ctx: QualifiedIdentifierContext): Any {
+        // Return the list of qualified identifiers for resolution by the containing element
+        val identifiers: MutableList<String?> = ArrayList()
```
issue: Some of the type inferences here are incorrect. The ids in this list aren't nullable.

antvaset replied:
Changed this to `val identifiers = ArrayList<String>()`

JPercival commented on Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/preprocessor/CqlPreprocessorElmCommonVisitor.kt:
```diff
@@ -0,0 +1,878 @@
+    }
+
+    override fun visit(tree: ParseTree): Any? {
+        Objects.requireNonNull(tree, "ParseTree required")
```
issue: Unneeded null check

JPercival commented on Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/preprocessor/CqlPreprocessorElmCommonVisitor.kt:
```diff
@@ -0,0 +1,878 @@
+        modelName: String?,
+        version: String?,
+        localIdentifier: String?
+    ): Model? {
```
question: This smells like an incorrect inference. All the parameters are nullable

JPercival commented on Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/preprocessor/LibraryInfo.kt:
```diff
@@ -0,0 +1,261 @@
+    var defaultUsingDefinition: UsingDefinitionInfo? = null
+        private set
+
+    private val usingDefinitions: MutableMap<String?, UsingDefinitionInfo>
```
issue: incorrect inference for these maps

antvaset replied:
Changed the initializers to
```private val usingDefinitions = LinkedHashMap<String, UsingDefinitionInfo>()
private val includeDefinitions = LinkedHashMap<String, IncludeDefinitionInfo>()
private val codesystemDefinitions = LinkedHashMap<String, CodesystemDefinitionInfo>()
private val valuesetDefinitions = LinkedHashMap<String, ValuesetDefinitionInfo>()
private val codeDefinitions = LinkedHashMap<String, CodeDefinitionInfo>()
private val conceptDefinitions = LinkedHashMap<String, ConceptDefinitionInfo>()
private val parameterDefinitions = LinkedHashMap<String, ParameterDefinitionInfo>()
private val expressionDefinitions = LinkedHashMap<String, ExpressionDefinitionInfo>()
private val functionDefinitions = LinkedHashMap<String, MutableList<FunctionDefinitionInfo>>()
private val contextDefinitions = ArrayList<ContextDefinitionInfo>()
private val definitions = HashMap<Interval, BaseInfo>()
```

antvaset commented:
I cleaned it up a bit more. As a rule of thumb:
* minimized the number of `?.` and `!!` used in the code
* removed all `@JvmField` (unless the property was used as a field _many_ times within existing `.java` files)
* removed types in variable and property declarations when they could be inferred
* preferred `parseString(ctx.termFromGrammar())!!` over `parseString(ctx.termFromGrammar())?.let { ... }` when we expected a non-nullable String based on the grammar.

JPercival commented:
@lukedegruchy reviewed a prior version of this PR and approved. Github didn't capture that.
